lib/alchemy/upgrader/four_point_zero.rb

```diff
@@ -1,3 +1,104 @@
+require 'thor'
+
+class Alchemy::Upgrader::FourPointZeroTasks < Thor
+  include Thor::Actions
+
+  no_tasks do
+
+    def convert_available_contents_to_nestable_elements
+      old_config_file = Rails.root.join('config', 'alchemy', 'elements.yml')
+      print "Reading config/alchemy/elements.yml ... "
+      elements_config = YAML.load_file(old_config_file)
+      puts "done."
+
+      print "Copying existing config file to config/alchemy/elements.yml.old ... "
+      FileUtils.copy old_config_file, Rails.root.join('config', 'alchemy', 'elements.yml.old')
+      puts "done."
+
+      print "Identifying `available_contents` ... "
+      elements_with_available_contents = elements_config.select { |e| e['available_contents'] }
+      puts 'done.'
+
+      print 'Converting to `nestable_elements` ... '
+      new_elements = elements_with_available_contents.inject([]) do |new_elements, old_element|
+        new_elements << modify_old_element(old_element.dup)
+        build_new_elements(old_element).inject(new_elements) do |ne, element_from_contents|
+          ne << element_from_contents
+        end
+      end.uniq.sort {|a, b| a['name'] <=> b['name'] }
+      puts 'done.'
+
+      print 'Writing new `config/alchemy/elements.yml` ... '
+      File.open(old_config_file, "w") do |f|
+        f.write new_elements.to_yaml
+      end
+      puts 'done.'
+
+      print 'Removing `render_new_content_link` helper from editor partials ... '
+      remove_new_content_link_from_editor_partials
+      puts 'done.'
+
+      print 'Adding hints for rendering nested elements to your views ... '
+      add_render_nested_elements_hints
+      puts 'done.'
+
+
+      puts "Please run `rails g alchemy:elements --skip` to generate partials for your new nested elements."
+    end
+  end
+
+  private
+
+  def modify_old_element(element)
+    nestable_elements = element['available_contents'].map do |content|
+      "addable_#{content['name']}"
+    end
+    element.delete('available_contents') # Hashes are mutable. Welcome!
+    element['nestable_elements'] = nestable_elements
+    element
+  end
+
+  def remove_new_content_link_from_editor_partials
+    editor_partials = Dir.glob(Rails.root.join('app', 'views', 'alchemy', 'elements', '*_editor*'))
+    system "sed -i '' '/^.*render_new_content_link.*$/d' #{editor_partials.join(' ')}"
+  end
+
+  def build_new_elements(element)
+    element['available_contents'].inject([]) do |collection, content|
+      collection << build_new_element(content)
+    end
+  end
+
+  def build_new_element(content)
+    {
+      'name' => "addable_#{content['name']}",
```

Comment:
I'm calling `uniq` on the collection of element definitions, which removes all of the ones that have the same name *and the same options*. I didn't know this, but it looks like `uniq` on a collection of Hashes actually deep-compares the Hashes, which is great here: All those that actually have differing configurations will be kept (sometimes under non-unique names), but the ones that are the same (which will be the vast majority of `available_contents` will be discarded. Shorter `element.yml` files for the win! (Sidenote: On the site I am converting right now, namespacing the addable elements with their source element just creates what I would call "noise definitions".) I mean: How many different ways can there be to render an `EssenceFile`? In case you have differing needs, you're always free to create more Elements manually, it's easy enough.